chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc

```diff
@@ -31,25 +33,213 @@ void MaybeClearAccessDeniedException(StorageArea* storage,
+    Page* page,
+    LocalDOMWindow* window) {
+  DCHECK(window);
+  if (!page)
+    return nullptr;
+
+  EphemeralStorageNamespaces* supplement =
+      Supplement<Page>::From<EphemeralStorageNamespaces>(page);
+  if (supplement)
+    return supplement;
+
+  Document* document = window->GetFrame()->GetDocument();
+  auto* web_frame = WebLocalFrameImpl::FromFrame(document->GetFrame());
+  WebViewImpl* webview = web_frame->ViewImpl();
+  if (!webview || !webview->Client())
+    return nullptr;
+  String session_storage_id =
+      String::FromUTF8(webview->Client()->GetSessionStorageNamespaceId()) +
+      String("/ephemeral-session-storage");
+
+  auto* security_origin =
+      page->MainFrame()->GetSecurityContext()->GetSecurityOrigin();
+  String domain = security_origin->RegistrableDomain();
+  // RegistrableDomain might return an empty string if this host is an IP
+  // address or a file URL.
+  if (domain.IsEmpty())
+    domain = String::FromUTF8(security_origin->ToUrlOrigin().Serialize());
+
+  String local_storage_id = domain + String("/ephemeral-local-storage");
+  supplement = MakeGarbageCollected<EphemeralStorageNamespaces>(
+      StorageController::GetInstance(), session_storage_id, local_storage_id);
+
+  ProvideTo(*page, supplement);
+  return supplement;
+}
+
+// static
+const char BraveDOMWindowStorage::kSupplementName[] = "BraveDOMWindowStorage";
+
+// static
+BraveDOMWindowStorage& BraveDOMWindowStorage::From(LocalDOMWindow& window) {
+  BraveDOMWindowStorage* supplement =
+      Supplement<LocalDOMWindow>::From<BraveDOMWindowStorage>(window);
+  if (!supplement) {
+    supplement = MakeGarbageCollected<BraveDOMWindowStorage>(window);
+    ProvideTo(window, supplement);
+  }
+  return *supplement;
 }
 
 // static
-StorageArea* BraveDOMWindowStorage::localStorage(LocalDOMWindow& window,
-                                            ExceptionState& exception_state) {
+StorageArea* BraveDOMWindowStorage::sessionStorage(
+    LocalDOMWindow& window,
+    ExceptionState& exception_state) {
+  return From(window).sessionStorage(exception_state);
+}
+
+// static
+StorageArea* BraveDOMWindowStorage::localStorage(
+    LocalDOMWindow& window,
+    ExceptionState& exception_state) {
+  return From(window).localStorage(exception_state);
+}
+
+BraveDOMWindowStorage::BraveDOMWindowStorage(LocalDOMWindow& window)
+    : Supplement<LocalDOMWindow>(window) {}
+
+StorageArea* BraveDOMWindowStorage::sessionStorage(
+    ExceptionState& exception_state) {
+  LocalDOMWindow* window = GetSupplementable();
   auto* storage =
-      DOMWindowStorage::From(window).localStorage(exception_state);
-  MaybeClearAccessDeniedException(storage, window, &exception_state);
+      DOMWindowStorage::From(*window).sessionStorage(exception_state);
+
+  MaybeClearAccessDeniedException(storage, *window, &exception_state);
+  if (base::FeatureList::IsEnabled(features::kBraveEphemeralStorage)) {
+    storage = ephemeralSessionStorage(storage);
```

Comment:
why not just unconditionally `return ephemeralSessionStorage(storage)` ? We have the check inside `ephemeralSessionStorage()`

Reply:
Okay. I'll fix this.